### DISCUSSION TITLE
Why does my program get stuck? Unexpected interaction between two tokio runtimes

### DISCUSSION THREAD

**Discussion by LoneMartian** [tokio-rs/tokio#6267] [Q&A] [answered]
Hi all, 

Async Rust beginner here. The sample code in my [repository](https://github.com/LoneMartian/sync-grpc-client) contains generated gRPC code for a very simple Greeter service. The macros in `src/lib.rs` are supposed to provide a blocking wrapper around a single-threaded tokio runtime. The test in `tests/integration_tests.rs` creates a new thread and runs the Greeter service on a second tokio runtime. Subsequently, the main thread instantiates a client and calls two functions on it.

Running `cargo test` never returns, as some of the futures appear to get stuck. Here a screenshot from `tokio-console`:

![Screenshot_20240104_085733](https://github.com/tokio-rs/tokio/assets/105127116/ef980875-a183-4437-b3b3-681004383cbf)

Some general observations:

- When refactoring client and server into separate binaries (e.g. `src/client.rs` and `src/server.rs`), the programs behave as expected.
- When using two `multi_thread` tokio runtimes, the integration test behaves as expected.
- When moving the server from the integration test (e.g. implementing it in Python), the test passes with the `current_thread` runtime.

I hope anyone can point out to me what's going on here.

**Comment by Darksonn** [comment]
Tasks on a current_thread runtime only run if you are actively calling `Runtime::block_on` on it. If you aren't doing that, then that's probably the cause.

**Reply by LoneMartian** [reply]
`block_on` is called [here](https://github.com/LoneMartian/sync-grpc-client/blob/master/src/lib.rs#L12) for the client, as well as [here](https://github.com/LoneMartian/sync-grpc-client/blob/master/tests/integration_tests.rs#L56) for the server.

**Reply by Darksonn** [reply, marked as answer]
Yes, but perhaps there are still tasks on the runtime when you return from `block_on`? Those tasks will stop running until you call `block_on` again.

**Reply by LoneMartian** [reply]
You seem to be correct indeed. If I call `block_on` again, e.g. in `drop`, the program behaves as expected. Now I need to understand this behavior. The [doc](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.block_on)  of `block_on` states the following:

> This runs the given future on the current thread, blocking until it is complete, and yielding its resolved result. Any tasks or timers which the future spawns internally will be executed on the runtime.

I guess the second sentence could explain my observation? Internally, new tasks are put onto the runtime, which are blocked from completing because the initial future has completed and I have returned from the `block_on` call.

**Reply by Darksonn** [reply]
Yes. A current_thread runtime does not spawn its own threads, so if you're not calling `block_on`, then it doesn't have anywhere to run the tasks.

**Reply by LoneMartian** [reply]
What still puzzles me is how the future that I'm waiting on (e.g. the RPC call) can complete (as stated by the `block_on` documentation) while the tasks it has spawned internally have not.

**Reply by Darksonn** [reply]
Yes, that does sound surprising. But it sounds like that's just how the library works.

**Reply by LoneMartian** [reply]
Another point of confusion is the observation that when I split server and client into separate binaries, both binaries are able to exit normally, even with current-thread runtimes. In the integration test case, the server seems to never catch the shutdown signal, as it is seemingly waiting for futures to complete. Do you have any explanation for this?

**Reply by LoneMartian** [reply]
I found an explanation: the runtime's `Drop` implementation drops all tasks, causing any open TCP streams to be closed, allowing the gRPC server to properly terminate. Thanks for the inputs.
